platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationMessageBuilder.java

```diff
@@ -392,23 +387,50 @@ String formatAdvice(List<String> replacements) {
         }
     }
 
-    static final Pattern ID_CLEAN_PATTERN = Pattern.compile("[A-Za-z]+");
+    public static final char DASH = '-';
 
     public static String createDefaultDeprecationId(String... ids) {
         StringBuilder sb = new StringBuilder();
         for (String id : ids) {
-            if(id == null){
+            if (id == null) {
                 continue;
             }
-            Matcher matcher = ID_CLEAN_PATTERN.matcher(id);
-            while (matcher.find()) {
-                sb.append(TextUtil.toLowerCaseLocaleSafe(matcher.group()));
-                sb.append("-");
+            CharSequence cleanId = getToken(id);
+            if (cleanId.length() > 0) {
+                sb.append(cleanId);
+                sb.append(DASH);
             }
         }
-        return removeEnd(sb.toString(), "-");
+        removeTrailingDashes(sb);
+        return sb.toString();
     }
```

Comment:
The string manipulation is sprinkled around in the method body, in `getToken` and in removeTrailingDashes(). I think it would be cleaner if `getToken()` would ensure that the result does not start nor finish with a dash. Then, this code can be simplified (please verify my suggestion; I just write it in the web review).
```
            CharSequence cleanId = getToken(id);
            String separator = "";
            if (cleanId.length() > 0) {
                sb.append(separator);
                sb.append(cleanId);
                separator = DASH;
            }
           // no need for removeTrailingDashes(sb);
```

Reply:
The `ids` collection should be first mapped with the `getToken(@Nullable String id)` function then the null values should be filtered.

Reply:
since the dash is also used as substitute for non ascii characters. There are situations where the logic would be more complicated than joining the parts.
I think removing the trailing dashes is effective also considering this is only a integer subtraction in the end, not an actual creation of a new array in the StringBuilder.
Also for the filtering of null values after the token extraction I'd need to check for null twice, inside the method and outside. 
This code is hard to get a lot more readable without losing efficiency to my understanding. the scattering of logic in functions keeps it cleaner because it explains what the lines are doing and is harder to ignore than a comment and therefore is intentional.